core/tests/ledger_cleanup.rs

```diff
@@ -0,0 +1,296 @@
+
+        println!(
+            "{},{},{},{},{},{},{},{},{},{},{}",
+            time_now
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap()
+                .as_millis(),
+            time_now.duration_since(*time_previous).unwrap().as_millis(),
+            start_slot,
+            batch_size,
+            entries,
+            max_slots,
+            storage_now,
+            storage_now as i64 - *storage_previous as i64,
+            cpu_user,
+            cpu_system,
+            cpu_idle,
+        );
+
+        *time_previous = time_now;
+        *storage_previous = storage_now;
+    }
+
+    #[test]
+    fn test_ledger_cleanup_compaction() {
+        let blocktree_path = get_tmp_ledger_path!();
+        let blocktree = Arc::new(Blocktree::open(&blocktree_path).unwrap());
+        let config = get_benchmark_config();
+        eprintln!("BENCHMARK CONFIG: {:?}", config);
+        eprintln!("LEDGER_PATH: {:?}", &blocktree_path);
+
+        let benchmark_slots = config.benchmark_slots;
+        let batch_size = config.batch_size;
+        let max_ledger_slots = config.max_ledger_slots;
+        let batches = benchmark_slots / batch_size;
+
+        let (sender, receiver) = channel();
+        let exit = Arc::new(AtomicBool::new(false));
+        let cleaner =
+            LedgerCleanupService::new(receiver, blocktree.clone(), max_ledger_slots, &exit);
+
+        let exit_cpu = Arc::new(AtomicBool::new(false));
+        let sys = CpuStatsUpdater::new(&exit_cpu);
+
+        let mut time_previous = SystemTime::now();
+        let mut storage_previous = 0;
+
+        emit_header();
+        emit_stats(
+            &mut time_previous,
+            &mut storage_previous,
+            0,
+            0,
+            0,
+            0,
+            &blocktree,
+            &sys.get_stats(),
+        );
+
+        for i in 0..batches {
+            let x = i * batch_size;
+
+            let (shreds, _) = make_many_slot_entries(x, batch_size, ENTRIES_PER_SLOT);
+            blocktree.insert_shreds(shreds, None, false).unwrap();
+            sender.send((i * batch_size, Pubkey::default())).unwrap();
+
+            emit_stats(
+                &mut time_previous,
+                &mut storage_previous,
+                x,
+                batch_size,
+                batch_size,
+                max_ledger_slots as i64,
+                &blocktree,
+                &sys.get_stats(),
+            );
+        }
+
+        let u1 = storage_previous;
+
+        sender.send((benchmark_slots, Pubkey::default())).unwrap();
+        emit_stats(
+            &mut time_previous,
+            &mut storage_previous,
+            benchmark_slots,
+            0,
+            0,
+            max_ledger_slots as i64,
+            &blocktree,
+            &sys.get_stats(),
+        );
+
+        std::thread::sleep(std::time::Duration::from_secs(6));
```

Comment:
What are these random sleeps?

Reply:
giving `rocksdb` subsystem time to flush

Reply:
(added constant ROCKSDB_FLUSH_GRACE_PERIOD_SECS for clarification)